sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregateHashMap.scala

```diff
@@ -0,0 +1,132 @@
+
+  def generate(): String = {
+    s"""
+       |public class $generatedClassName {
+       |${initializeAggregateHashMap()}
+       |
+       |${generateFindOrInsert()}
+       |
+       |${generateEquals()}
+       |
+       |${generateHashFunction()}
+       |}
+     """.stripMargin
+  }
+
+  def initializeAggregateHashMap(): String = {
+    val generatedSchema: String =
+      s"""
+         |new org.apache.spark.sql.types.StructType()
+         |${(groupingKeySchema ++ bufferSchema).map(key =>
+            s""".add("${key.name}", org.apache.spark.sql.types.DataTypes.${key.dataType})""")
+            .mkString("\n")};
+       """.stripMargin
+
+    s"""
+       |  private org.apache.spark.sql.execution.vectorized.ColumnarBatch batch;
+       |  private int[] buckets;
+       |  private int numBuckets;
+       |  private int maxSteps;
+       |  private int numRows = 0;
+       |  private org.apache.spark.sql.types.StructType schema = $generatedSchema
+       |
+       |  public $generatedClassName(int capacity, double loadFactor, int maxSteps) {
+       |    assert (capacity > 0 && ((capacity & (capacity - 1)) == 0));
+       |    this.maxSteps = maxSteps;
+       |    numBuckets = (int) (capacity / loadFactor);
+       |    batch = org.apache.spark.sql.execution.vectorized.ColumnarBatch.allocate(schema,
+       |      org.apache.spark.memory.MemoryMode.ON_HEAP, capacity);
+       |    buckets = new int[numBuckets];
+       |    java.util.Arrays.fill(buckets, -1);
+       |  }
+       |
+       |  public $generatedClassName() {
+       |    new $generatedClassName(1 << 16, 0.25, 5);
+       |  }
+     """.stripMargin
+  }
+
+  def generateHashFunction(): String = {
+    s"""
+       |// TODO: Improve this Hash Function
+       |private long hash($groupingKeySignature) {
+       |  return ${groupingKeys.map(_._2).mkString(" & ")};
+       |}
+     """.stripMargin
+  }
+
+  def generateEquals(): String = {
+    s"""
+       |private boolean equals(int idx, $groupingKeySignature) {
+       |  return ${groupingKeys.zipWithIndex.map(key =>
+            s"batch.column(${key._2}).getLong(buckets[idx]) == ${key._1._2}").mkString(" && ")};
+       |}
+     """.stripMargin
+  }
+
+  def generateFindOrInsert(): String = {
+    s"""
+       |public org.apache.spark.sql.execution.vectorized.ColumnarBatch.Row findOrInsert(${
+          groupingKeySignature}) {
+       |  int idx = find(${groupingKeys.map(_._2).mkString(", ")});
+       |  if (idx != -1 && buckets[idx] == -1) {
+       |    ${groupingKeys.zipWithIndex.map(key =>
+              s"batch.column(${key._2}).putLong(numRows, ${key._1._2});").mkString("\n")}
+       |    ${bufferValues.zipWithIndex.map(key =>
+              s"batch.column(${groupingKeys.length + key._2}).putLong(numRows, 0);")
+              .mkString("\n")}
+       |    buckets[idx] = numRows++;
+       |  }
+       |  return batch.getRow(buckets[idx]);
+       |}
+       |
+       |private int find($groupingKeySignature) {
```

Comment:
Let's simplify this. The generated code only needs findOrInsert() and doesn't need find.